fendermint/vm/topdown/proof-service/src/service.rs

```diff
@@ -0,0 +1,281 @@
+        let max_instance = last_committed + self.config.lookahead_instances;
+
+        tracing::debug!(
+            last_committed,
+            next_instance,
+            max_instance,
+            "Checking for new F3 certificates"
+        );
+
+        // Process instances IN ORDER - this is critical for F3
+        for instance_id in next_instance..=max_instance {
+            // Skip if already cached
+            if self.cache.contains(instance_id) {
+                tracing::debug!(instance_id, "Proof already cached");
+                continue;
+            }
+
+            // ====================
+            // STEP 1: FETCH + VALIDATE certificate (single operation!)
+            // ====================
+            let validated = match self.f3_client.fetch_and_validate(instance_id).await {
+                Ok(cert) => cert,
+                Err(e)
+                    if e.to_string().contains("not found")
+                        || e.to_string().contains("not available") =>
+                {
+                    // Certificate not available yet - STOP HERE!
+                    // Don't try higher instances as they depend on this one
+                    tracing::debug!(instance_id, "Certificate not available, stopping lookahead");
+                    break;
+                }
+                Err(e) => {
+                    return Err(e).with_context(|| {
+                        format!(
+                            "Failed to fetch and validate certificate for instance {}",
+                            instance_id
+                        )
+                    });
+                }
+            };
+
+            tracing::info!(
+                instance_id,
+                ec_chain_len = validated.f3_cert.ec_chain.suffix().len(),
+                "Certificate fetched and validated successfully"
+            );
+
+            // ====================
+            // STEP 2: GENERATE proof bundle
+            // ====================
+            let proof_bundle = self
+                .generate_proof_for_certificate(&validated.f3_cert)
+                .await
+                .context("Failed to generate proof bundle")?;
+
+            // ====================
+            // STEP 3: CACHE the result
+            // ====================
+            let entry = CacheEntry::new(
+                &validated.f3_cert,
+                proof_bundle,
+                "F3 RPC".to_string(), // source_rpc
+            );
+
+            self.cache.insert(entry)?;
+
+            tracing::info!(
+                instance_id,
+                "Successfully cached validated certificate and proof bundle"
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Generate proof bundle for a specific certificate
+    ///
+    /// Extracts the highest epoch, fetches tipsets, and generates proofs.
+    async fn generate_proof_for_certificate(
+        &self,
+        f3_cert: &filecoin_f3_certs::FinalityCertificate,
```

Comment:
Certificates should be disposed after being processed. The latest tipset should be derived from the light client state.